File: MQTT_main_server/Redis_Lcard/Utils/RespMessage/mqttrespmessage.cpp
#include "mqttrespmessage.h"

using json = nlohmann::json;

MQTTRespMessage::MQTTRespMessage(const std::string &topic,
                                 IResponse& response,
                                 const std::string &device,
                                 const std::string &time):
    MQTTMessage(topic, device, time),
    response(response){}

json MQTTRespMessage::toJSON() const {
    json message = MQTTMessage::toJSON();
    message["response"] = response.toString();
    message["time"] = time;
    return message;
}

std::string MQTTRespMessage::toString() const {
    std::ostringstream oss;
    oss << "MQTTMessage(topic=" << topic;
    oss << ", Response = {" << response.toString();
    oss << "}, device=" << device
        << ", time=" << time
        << ")";

    return oss.str();
}

File: MQTT_main_server/Utils/MQTTRespMessage.py
from Utils.MQTTMessage import MQTTMessage
import time
import json

class MQTTRespMessage(MQTTMessage):
    def __init__(self, topic, device, response : json, time = time.time()):
        super().__init__(topic, device, time)
        self.response : json = response

    def to_json(self):
        """Возвращает сообщение в формате JSON."""
        message = {
            "topic": self.topic,
            "device": self.device,
            "response": json.dumps(self.response),
            "time": self.time
        }
        message = json.dumps(message)
        return message

    @staticmethod
    def from_json(json_data):
        """Создает объект MQTTMessage из JSON строки."""
        try:
            data = json.loads(json_data)
            return MQTTRespMessage(
                topic=data.get("topic"),
                device=data.get("device"),
                response=json.loads(data.get("response")),
                time=data.get("time")
            )
        except json.JSONDecodeError as e:
            raise ValueError(f"Ошибка декодирования JSON: {e}")

    def __str__(self):
        """Возвращает строковое представление сообщения."""
        return (f"MQTTMessage(topic={self.topic}, "
                f"device={self.device}, "
                f"responce={self.response}, time={self.time}")
